pkg/server/settingsworker_test.go

```diff
@@ -293,3 +304,164 @@ func TestSettingsShowAll(t *testing.T) {
+			return fmt.Errorf(`did not observe spans in batches of size 1`)
+		}
+		return nil
+	})
+}
+
+// TestSettingsTraceSamplingRate tests that setting the sampling rate
+// of traces will correctly limit the percentage of traces sampled.
+func TestSettingsTraceSamplingRate(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	// Start test cluster containing a test db and test table
+	ctx := context.Background()
+	tc := testcluster.StartTestCluster(t, 3, base.TestClusterArgs{})
+	defer tc.Stopper().Stop(ctx)
+
+	execStatement := func(statement string) {
+		if _, err := tc.ServerConn(rand.Intn(3)).Exec(statement); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	execStatement(`CREATE DATABASE t`)
+	execStatement(`CREATE TABLE test (k INT PRIMARY KEY, v INT)`)
+
+	// Set up tracing collector and variable to count the number of traces
+	// collected from INSERT INTO statements
+	nTraces := int64(0)
+	collectorHandlerFcn := func(rw http.ResponseWriter, req *http.Request) {
+		bodyBytes, err := ioutil.ReadAll(req.Body)
+		if err != nil {
+			return
+		}
+		t := thrift.NewTMemoryBuffer()
+		if _, err = t.Write(bodyBytes); err != nil {
+			return
+		}
+		p := thrift.NewTBinaryProtocolTransport(t)
+		_, batchSize, err := p.ReadListBegin()
+		if err != nil {
+			return
+		}
+
+		for i := 0; i < batchSize; i += 1 {
+			span := zipkincore.Span{}
+			if err = span.Read(p); err != nil {
+				return
+			}
+
+			if span.Name == `exec stmt` {
+				for _, annotation := range span.Annotations {
+					if strings.Contains(annotation.Value, `INSERT INTO test VALUES`) &&
+						strings.Contains(annotation.Value, `in state: Open`) {
+						atomic.AddInt64(&nTraces, 1)
+					}
+				}
+			}
+		}
+	}
+	traceCollectorServer := httptest.NewServer(http.HandlerFunc(collectorHandlerFcn))
+	defer traceCollectorServer.Close()
+
+	// Configure tracing in the test cluster
+	traceCollectorServerURL, err := url.Parse(traceCollectorServer.URL)
+	if err != nil {
+		t.Fatal(err)
+	}
+	setClusterTraceSamplingRate := func(rate float64) {
+		execStatement(fmt.Sprintf(`SET CLUSTER SETTING trace.zipkin.sampling.rate = %f`, rate))
+	}
+	execStatement(fmt.Sprintf(`SET CLUSTER SETTING trace.zipkin.collector = '%s'`, traceCollectorServerURL.Host))
+	execStatement(`SET CLUSTER SETTING trace.zipkin.collector.batch.size = 200`)
+
+	// Function that writes 1000 rows
+	startKey := 0
+	write1000Rows := func() {
+		waitGroup := sync.WaitGroup{}
+		for key := startKey; key < startKey+1000; key += 10 {
+			waitGroup.Add(1)
+			go func(wg *sync.WaitGroup, k int) {
+				defer wg.Done()
+				for v := 0; v < 10; v++ {
+					execStatement(fmt.Sprintf(`INSERT INTO test VALUES (%d, %d)`, k+v, v))
+				}
+			}(&waitGroup, key)
+		}
+		waitGroup.Wait()
+		startKey = startKey + 1000
+	}
+
+	// Test 100% Sampling Rate
+	setClusterTraceSamplingRate(1.0)
+	nTraces = 0
+	write1000Rows()
+	testutils.SucceedsSoon(t, func() error {
+		currentTraces := atomic.LoadInt64(&nTraces)
+		if currentTraces < 950 && currentTraces > 1050 {
```

Comment:
Locally, I was able to get exactly 1000. In CI, this failed a few times. When I got less than 1000, I could see in the logs that the collector buffer would get filled and spans would get dropped. Sometimes, I would actually get over 1000, which I assume is due to retrying. 

If we really wanted exactly 1000 here, I could use map to track & count the unique keys being inserted. However, I don't think it is necessary because the main purpose of the test is to see if we can reduce the number of traces by setting a lower sampling rate. Checking to see that if we get roughly 1000 with a rate of 1.0 and then roughly 500 with a rate of 0.5 should be enough to test this.